CMSgov/dpc-app
# DPC-1302: Replicate in implementer portal - Provider Organization Management page

### Fixes [DPC-1302](https://jira.cms.gov/browse/DPC-1302)


The Implementer app will make calls to the Golang API for provider organization with org details and the ability to create a system with the provider org's initial token. Users will then be able to add public key/signature pairs and create new client tokens as needed and view a list of added public keys and tokens that have been added/created.

### Proposed Changes



- [x] User can view provider org page w/ provider org info
- [x] User can create system for provider org
- [x] User can create public key
- [x] User can view list of all public key ids
- [x] User can delete public key
- [x] User can create client token
- [x] User can view client token
- [x] User can view list of client tokens
- [x] User can delete client token

### Change Details


- Added `omitempty` to `ips` in `ProxyCreateSystemResponse`
- How to create public key and signature instructions
- Add `rails-controller-testing` for testing purposes
- Add how to create public key/signature pair instructions to public key form
  - Add `copy-text.js` to copy openssl commands
  - Add relevant accordion styling
  - Add `snippet.txt` with update text (`This is the snippet used to verify a key pair.`) for download to create signature.
- Styling for public keys and credentials
- Add `imp_id`, `api_service`, `missing_key`, and `render_error` methods to application controller
- `dpc-impl/app/services/api_client.rb` lines 113 - 119: Because the `create_client_token` returns the client token as a string had to add a check for the response body that if a JSON object continue with `eval()` otherwise return response body as is.
- Add Client Token Manager to help manage http calls for client token creation and deletion
- Add Public Key Manager to help manage http calls for system creations and public key creation and deletion.
  - Put in checks for valid encoding and signature presence

### Security Implications



- [ ] new software dependencies



- [ ] security controls or supporting software altered



- [ ] new data stored or transmitted



- [ ] security checklist is completed for this change



- [ ] requires more information or team discussion to evaluate security implications



- [x] no PHI/PII is affected by this change



### Acceptance Validation



Due to time constraints with the upcoming code freeze on August 5, I created [DPC-1638](https://jira.cms.gov/browse/DPC-1638) to address missing tests for the provider org show page, client token management, and public key management. This will allow the Implementer App to be able to be released on Aug 12.

I also created [DPC-1639](https://jira.cms.gov/browse/DPC-1639) to add in the 25 character limit to client token labels and the GUI that already exists in the web app.

Below are screen captures of workflows to prove their efficacy:


**View Provider Org Page**\
![recording](https://user-images.githubusercontent.com/44708048/128189766-892255be-82d8-4747-9d4d-b40721bcd8e7.gif)

**Create System & View Initial Token**\
![recording (1)](https://user-images.githubusercontent.com/44708048/128190027-675961fd-5eb0-4999-b8d1-45c8431872e8.gif)

**Adding New Public Key**\
![recording](https://user-images.githubusercontent.com/44708048/128207548-dd78bd47-baa7-4aaf-92a4-0495e6638f0e.gif)

**Deleting Public Key**\
![recording](https://user-images.githubusercontent.com/44708048/128207906-f00f5f08-df5f-465b-8a5c-ecd2fadd7be8.gif)

**Public Key Errors**\
**Cannot submit blank form or upload Private key**\
![recording](https://user-images.githubusercontent.com/44708048/128209056-bcf31087-cec5-4a89-a483-1d394d9d9195.gif)

**Add New Client Token**\
![recording](https://user-images.githubusercontent.com/44708048/128209233-e326e43b-b87b-4d56-8759-9fa6bfa8a9b8.gif)

**Delete Client Token**\
![recording](https://user-images.githubusercontent.com/44708048/128209788-ad3d007e-2a07-49db-8277-debb63bdcef4.gif)

**Client Token Must Have Label**\
![recording](https://user-images.githubusercontent.com/44708048/128209929-e0357330-de39-4a26-b89c-bf49875c6afd.gif)

### Feedback Requested
